### PR TITLE
Histogram Rollout - Algorithms 108-112

### LINKED ISSUE #17799: Histogram Rollout - 108-112
HistogramData rollout: algorithms 108-112

This issue is part of the HistogramData rollout, see https://github.com/mantidproject/mantid/issues/17641.
Algorithms covered in this issue are:

```
Framework/DataHandling/src/LoadFITS.cpp
Framework/DataHandling/src/LoadGSS.cpp
Framework/DataHandling/src/LoadILLIndirect.cpp
Framework/DataHandling/src/LoadILLReflectometry.cpp
Framework/DataHandling/src/LoadILLSANS.cpp
```

### PR DESCRIPTION
**Pull request by DavidFair**
This PR must be merged **after** #17788 

HistogramData rollout: algorithms 108-112
## Description of work.

This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:

```
Framework/DataHandling/src/LoadFITS.cpp
Framework/DataHandling/src/LoadGSS.cpp
Framework/DataHandling/src/LoadILLIndirect.cpp
Framework/DataHandling/src/LoadILLReflectometry.cpp
Framework/DataHandling/src/LoadILLSANS.cpp
```

Performance tests were added for:

```
LoadGSS
LoadILLIndirect
```

The following algorithms have improved performance:

```
LoadGSS - 20%
```
## To test.

Code review.

Fixes #17799 .

Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/WienerSmooth.cpp [hunk 1]
```diff
@@ -365,13 +366,16 @@ WienerSmooth::smoothSingleSpectrum(API::MatrixWorkspace_sptr inputWS,
 
   // copy the x-values and errors from the original spectrum
   // remove the last values for odd-sized inputs
+
+  out->setSharedX(0, inputWS->sharedX(wsIndex));
+  out->setSharedE(0, inputWS->sharedE(wsIndex));
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/WienerSmooth.cpp]: Keep this inside the `if` statement? The overhead may be very small, but it would also be clearer what the intention is. Also I am not sure if this would actually work. It might throw a size mismatch error for `ifOddSize == true`.

File: Framework/Algorithms/src/WienerSmooth.cpp [hunk 2]
```diff
@@ -385,8 +389,9 @@ WienerSmooth::smoothSingleSpectrum(API::MatrixWorkspace_sptr inputWS,
  *   centres are used.
  * @return :: A pair of start x and end x.
  */
-std::pair<double, double> WienerSmooth::getStartEnd(const MantidVec &X,
-                                                    bool isHistogram) const {
+std::pair<double, double>
+WienerSmooth::getStartEnd(const Mantid::HistogramData::HistogramX &X,
+                          bool isHistogram) const {
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/WienerSmooth.cpp]: This could have been made nicer by passing `Points`. The `isHistogram` flag would then not be necessary. However, that would add significant overhead since only two of the points are needed, so I agree that this should be kept as it is 👍 .

File: Framework/DataHandling/src/LoadFITS.cpp [hunk 3]
```diff
@@ -860,10 +860,10 @@ void LoadFITS::readDataToWorkspace(const FITSInfo &fileInfo, double cmpp,
...
   PARALLEL_FOR_NO_WSP_CHECK()
   for (int i = 0; i < static_cast<int>(nrows); ++i) {
-    auto &dataX = ws->dataX(i);
-    auto &dataY = ws->dataY(i);
-    auto &dataE = ws->dataE(i);
-    std::fill(dataX.begin(), dataX.end(), static_cast<double>(i) * cmpp);
+    auto &xVals = ws->mutableX(i);
+    auto &yVals = ws->mutableY(i);
+    auto &eVals = ws->mutableE(i);
+    std::fill(xVals.begin(), xVals.end(), static_cast<double>(i) * cmpp);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadFITS.cpp]: Just use `operator=`?

File: Framework/DataHandling/src/LoadGSS.cpp [hunk 4]
```diff
@@ -253,9 +262,9 @@ API::MatrixWorkspace_sptr LoadGSS::loadGSASFile(const std::string &filename,
         std::vector<double> storeY = vecY;
         std::vector<double> storeE = vecE;
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadGSS.cpp]: Do we still need this copy?

File: Framework/DataHandling/src/LoadGSS.cpp [hunk 5]
```diff
@@ -253,9 +262,9 @@ API::MatrixWorkspace_sptr LoadGSS::loadGSASFile(const std::string &filename,
...
 
-        gsasDataX.push_back(storeX);
-        gsasDataY.push_back(storeY);
-        gsasDataE.push_back(storeE);
+        gsasDataX.emplace_back(storeX);
+        gsasDataY.emplace_back(storeY);
+        gsasDataE.emplace_back(storeE);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadGSS.cpp]: Use `std::move`?

Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadGSS.cpp]: If you use move here, it might pay of the complement it with a `reserve` call for `vecX`, `vecY`, and `vecE`, to speed up the `push_backs` done below.

File: Framework/DataHandling/src/LoadGSS.cpp [hunk 6]
```diff
@@ -393,9 +402,9 @@ API::MatrixWorkspace_sptr LoadGSS::loadGSASFile(const std::string &filename,
 
   // Push the vectors (X, Y, E) of the last bank to gsasData
   if (!vecX.empty()) { // Put final spectra into data
-    gsasDataX.push_back(vecX);
-    gsasDataY.push_back(vecY);
-    gsasDataE.push_back(vecE);
+    gsasDataX.emplace_back(vecX);
+    gsasDataY.emplace_back(vecY);
+    gsasDataE.emplace_back(vecE);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadGSS.cpp]: `std::move`?

File: Framework/DataHandling/src/LoadILLIndirect.cpp [hunk 7]
```diff
@@ -250,21 +249,20 @@ void LoadILLIndirect::loadDataIntoTheWorkSpace(
   Progress progress(this, 0, 1, m_numberOfTubes * m_numberOfPixelsPerTube +
                                     nb_monitors + nb_SD_detectors);
 
-  // Assign fake values to first X axis <<to be completed>>
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLIndirect.cpp]: Preserve this comment?

File: Framework/DataHandling/src/LoadILLIndirect.cpp [hunk 8]
```diff
@@ -250,21 +249,20 @@ void LoadILLIndirect::loadDataIntoTheWorkSpace(
...
-  for (size_t i = 0; i <= m_numberOfChannels; ++i) {
-    m_localWorkspace->dataX(0)[i] = double(i);
-  }
+  std::vector<double> increasingVals(m_numberOfChannels);
+  std::iota(increasingVals.begin(), increasingVals.end(), 1);
+  m_localWorkspace->mutableX(0) = increasingVals;
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLIndirect.cpp]:
```cpp
#include "MantidHistogramData/LinearGenerator.h"

m_localWorkspace->setBinEdges(0, m_numberOfChannels + 1, LinearGenerator(1.0, 1.0));
```

File: Framework/DataHandling/src/LoadILLIndirect.cpp [hunk 9]
```diff
@@ -250,21 +249,20 @@ void LoadILLIndirect::loadDataIntoTheWorkSpace(
...
 
   // First, Monitor
   for (size_t im = 0; im < nb_monitors; im++) {
 
     if (im > 0) {
-      m_localWorkspace->dataX(im) = m_localWorkspace->readX(0);
+      m_localWorkspace->setSharedX(im, m_localWorkspace->sharedX(0));
     }
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLIndirect.cpp]: You could actually remove the `if`, since `setSharedX` handles self-assignment correctly.

File: Framework/DataHandling/src/LoadILLIndirect.cpp [hunk 10]
```diff
@@ -149,7 +150,7 @@ void LoadILLIndirect::loadDataDetails(NeXus::NXEntry &entry) {
 
   m_numberOfTubes = static_cast<size_t>(data.dim0());
   m_numberOfPixelsPerTube = static_cast<size_t>(data.dim1());
-  m_numberOfChannels = static_cast<size_t>(data.dim2());
+  m_numberOfChannels = static_cast<size_t>(data.dim2()) + 1; // +1 For bin edges
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLIndirect.cpp]: Why?

File: Framework/DataHandling/src/LoadILLIndirect.cpp [hunk 11]
```diff
@@ -209,7 +210,7 @@ void LoadILLIndirect::initWorkSpace(
   m_localWorkspace = WorkspaceFactory::Instance().create(
       "Workspace2D",
       m_numberOfHistograms + monitorsData.size() + m_numberOfSimpleDetectors,
-      m_numberOfChannels + 1, m_numberOfChannels);
+      m_numberOfChannels, m_numberOfChannels);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLIndirect.cpp]: If you have bin edges in a workspace the first of these arguments should have the `+1`.

File: Framework/DataHandling/src/LoadILLIndirect.cpp [hunk 12]
```diff
@@ -240,8 +241,6 @@ void LoadILLIndirect::loadDataIntoTheWorkSpace(
   dataSD.load();
 
   // Assign calculated bins to first X axis
-  ////  m_localWorkspace->dataX(0).assign(detectorTofBins.begin(),
-  /// detectorTofBins.end());
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLIndirect.cpp]: Also remove the comment?

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 13]
```diff
@@ -286,9 +286,6 @@ void LoadILLReflectometry::loadDataIntoTheWorkSpace(
   data.load();
 
   // Assign calculated bins to first X axis
-  ////  m_localWorkspace->dataX(0).assign(detectorTofBins.begin(),
-  /// detectorTofBins.end());
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: Also remove the comment?

File: Framework/DataHandling/src/LoadILLSANS.cpp [hunk 14]
```diff
@@ -256,12 +256,13 @@ LoadILLSANS::loadDataIntoWorkspaceFromMonitors(NeXus::NXEntry &firstEntry,
         positionsBinning.push_back(static_cast<double>(i));
 
       // Assign X
-      m_localWorkspace->dataX(firstIndex)
+      m_localWorkspace->mutableX(firstIndex)
           .assign(positionsBinning.begin(), positionsBinning.end());
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLSANS.cpp]: `m_localWorkspace->setBinEdges(firstIndex, vectorSize, LinearGenerator(0.0, 1.0));` and remove `positionsBinning`?

File: Framework/DataHandling/src/LoadILLSANS.cpp [hunk 15]
```diff
@@ -256,12 +256,13 @@ LoadILLSANS::loadDataIntoWorkspaceFromMonitors(NeXus::NXEntry &firstEntry,
...
       // Assign Y
-      m_localWorkspace->dataY(firstIndex).assign(data(), data() + data.dim2());
+      m_localWorkspace->mutableY(firstIndex)
+          .assign(data(), data() + data.dim2());
       // Assign Error
-      MantidVec &E = m_localWorkspace->dataE(firstIndex);
+      auto &E = m_localWorkspace->mutableE(firstIndex);
       std::transform(data(), data() + data.dim2(), E.begin(),
                      LoadHelper::calculateStandardError);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLSANS.cpp]: Is this just computing the `sqrt` of the counts? If so, this could be done internally automatically by `Histogram`:

```cpp
m_localWorkspace->Histogram(firstIndex, BinEdges(vectorSize, LinearGenerator(0.0, 1.0)), Counts(data(), data() + data.dim2()); // NOTE: no E specified, so it defaults to sqrt(Counts)
```

File: Framework/DataHandling/src/LoadILLSANS.cpp (hunk 14 above)
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLSANS.cpp]: (see below first)

File: Framework/DataHandling/src/LoadILLSANS.cpp [hunk 16]
```diff
@@ -303,23 +304,25 @@ size_t LoadILLSANS::loadDataIntoWorkspaceFromHorizontalTubes(
...
       // Assign Y
       int *data_p = &data(static_cast<int>(j), static_cast<int>(i), 0);
-      m_localWorkspace->dataY(spec).assign(data_p, data_p + data.dim2());
+      m_localWorkspace->mutableY(spec).assign(data_p, data_p + data.dim2());
 
       // Assign Error
-      MantidVec &E = m_localWorkspace->dataE(spec);
+      auto &E = m_localWorkspace->mutableE(spec);
       std::transform(data_p, data_p + data.dim2(), E.begin(),
                      LoadHelper::calculateStandardError);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLSANS.cpp]: See comment above, maybe the whole block can be written with a single `setHistogram` call.

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 17]
```diff
@@ -396,15 +394,16 @@ void LoadILLReflectometry::loadDataIntoTheWorkSpace(
...
       int *data_p = &data(static_cast<int>(i), static_cast<int>(j), 0);
-      m_localWorkspace->dataY(spec + nb_monitors)
+      m_localWorkspace->mutableY(spec + nb_monitors)
           .assign(data_p, data_p + m_numberOfChannels);
 
       // Assign Error
-      MantidVec &E = m_localWorkspace->dataE(spec + nb_monitors);
+      auto &E = m_localWorkspace->mutableE(spec + nb_monitors);
       std::transform(data_p, data_p + m_numberOfChannels, E.begin(),
                      LoadHelper::calculateStandardError);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: See below, if this is just computing the `sqrt` of the counts then you can rewrite the whole block with a single `setHistogram` call.

File: Framework/DataHandling/src/LoadILLSANS.cpp [hunk 18]
```diff
@@ -355,23 +358,25 @@ size_t LoadILLSANS::loadDataIntoWorkspaceFromVerticalTubes(
...
       // Assign Y
       int *data_p = &data(static_cast<int>(i), static_cast<int>(j), 0);
-      m_localWorkspace->dataY(spec).assign(data_p, data_p + data.dim2());
+      m_localWorkspace->mutableY(spec).assign(data_p, data_p + data.dim2());
 
       // Assign Error
-      MantidVec &E = m_localWorkspace->dataE(spec);
+      auto &E = m_localWorkspace->mutableE(spec);
       std::transform(data_p, data_p + data.dim2(), E.begin(),
                      LoadHelper::calculateStandardError);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLSANS.cpp]: See comments above regarding `setHistogram`.

**Review by SimonHeybrock — COMMENTED** [review]
Looks mostly good, some comments below:

File: Framework/DataHandling/src/LoadILLIndirect.cpp (hunk 10 above)
Comment (DavidFair) [inline Framework/DataHandling/src/LoadILLIndirect.cpp]: Whenever `m_numberOfChannels` was used throughout the code one was always being appended to it so I moved it to assignment instead

File: Framework/DataHandling/src/LoadGSS.cpp (hunk 5 above)
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadGSS.cpp]: Right, I did not notice that the size was unknown.

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 19]
```diff
@@ -389,21 +389,10 @@ void LoadILLReflectometry::loadDataIntoTheWorkSpace(
...
-      m_localWorkspace->setHistogram((spec + nb_monitors), std::move(histo));
-
-      ++spec;
-      progress.report();
-    }
-  }
+  const HistogramData::Counts histoCounts(data(), data() + m_numberOfChannels);
+  const HistogramData::Histogram histo(std::move(binEdges),
+                                       std::move(histoCounts));
+  m_localWorkspace->setHistogram((spec + nb_monitors), std::move(histo));
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: `setHistogram` is templated and forwards arguments to a `Histogram` constructor. That is you could directly pass the bin edges and counts to `setHistogram`, no need for the helper variable `histo`.

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 20]
```diff
@@ -368,50 +364,33 @@ void LoadILLReflectometry::loadDataIntoTheWorkSpace(
...
-      std::transform(data_p, data_p + m_numberOfChannels, E.begin(),
-                     LoadHelper::calculateStandardError);
-
-      ++spec;
-      progress.report();
-    }
-  } // for m_numberOfTubes
+  const HistogramData::Counts histoCounts(data(), data() + m_numberOfChannels);
+  m_localWorkspace->setHistogram((spec + nb_monitors), std::move(binEdges),
+                                 std::move(histoCounts));
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: Should this still be in the loop over `i`, only the `j` loop is gone?

File: Framework/DataHandling/src/LoadILLSANS.cpp [hunk 21]
```diff
@@ -298,35 +293,19 @@ size_t LoadILLSANS::loadDataIntoWorkspaceFromHorizontalTubes(
...
-                     LoadHelper::calculateStandardError);
+  const HistogramData::BinEdges binEdges(timeBinning);
+  const HistogramData::Counts histoCounts(data(), data() + data.dim2());
 
-      ++spec;
-      progress.report();
-    }
-  }
+  m_localWorkspace->setHistogram(firstIndex, std::move(binEdges),
+                                 std::move(histoCounts));
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLSANS.cpp]: Same here?

File: Framework/DataHandling/src/LoadILLReflectometry.cpp [hunk 22]
```diff
@@ -389,8 +389,11 @@ void LoadILLReflectometry::loadDataIntoTheWorkSpace(
   }
 
   const HistogramData::Counts histoCounts(data(), data() + m_numberOfChannels);
-  m_localWorkspace->setHistogram((spec + nb_monitors), std::move(binEdges),
-                                 std::move(histoCounts));
+
+  for (size_t i = 0; i < m_numberOfTubes; ++i) {
+    m_localWorkspace->setHistogram((spec + nb_monitors), binEdges, histoCounts);
+    spec += m_numberOfPixelsPerTube;
+  }
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLReflectometry.cpp]: Looking at the code again I think you actually need both loops. Each pixel in a tube is a separate spectrum. Sorry for the confusion.

File: Framework/Algorithms/src/UnwrapMonitor.cpp [hunk 23]
```diff
@@ -109,9 +109,9 @@ void UnwrapMonitor::exec() {
       // If the detector flightpath is missing, zero the data
       g_log.debug() << "Detector information for workspace index " << i
                     << " is not available.\n";
-      tempWS->mutableX(i) = 0.0;
-      tempWS->mutableY(i) = 0.0;
-      tempWS->mutableE(i) = 0.0;
+      tempWS->mutableX(i).assign(tempWS->x(i).size(), 0.0);
+      tempWS->mutableY(i).assign(tempWS->y(i).size(), 0.0);
+      tempWS->mutableE(i).assign(tempWS->e(i).size(), 0.0);
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/UnwrapMonitor.cpp]: Why?

File: Framework/Algorithms/src/UnwrapMonitor.cpp [hunk 24]
```diff
@@ -142,7 +142,7 @@ void UnwrapMonitor::exec() {
     // Get the maximum number of bins (excluding monitors) for the rebinning
     // below
     if (!spectrumInfo.isMonitor(i)) {
-      const int XLen = static_cast<int>(tempWS->x(i).size());
+      const size_t XLen = static_cast<int>(tempWS->x(i).size());
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/UnwrapMonitor.cpp]: ?

File: Framework/Algorithms/src/WienerSmooth.cpp [hunk 25]
```diff
@@ -369,9 +369,10 @@ WienerSmooth::smoothSingleSpectrum(API::MatrixWorkspace_sptr inputWS,
 
   if (isOddSize) {
     auto histogram = out->histogram(0);
+    auto newSize = histogram.y().size() - 1;
+
     histogram.setSharedX(inputWS->sharedX(wsIndex));
     histogram.setSharedE(inputWS->sharedE(wsIndex));
-    auto newSize = histogram.y().size() - 1;
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/WienerSmooth.cpp]: Why this change? Declaring variables close to where they are used is usually considered better.

File: Framework/DataHandling/src/LoadILLIndirect.cpp [hunk 26]
```diff
@@ -275,15 +270,16 @@ void LoadILLIndirect::loadDataIntoTheWorkSpace(
...
       int *data_p = &data(static_cast<int>(i), static_cast<int>(j), 0);
-      m_localWorkspace->dataY(spec + nb_monitors)
+      m_localWorkspace->mutableY(spec + nb_monitors)
           .assign(data_p, data_p + m_numberOfChannels);
 
       // Assign Error
-      MantidVec &E = m_localWorkspace->dataE(spec + nb_monitors);
+      auto &E = m_localWorkspace->mutableE(spec + nb_monitors);
       std::transform(data_p, data_p + m_numberOfChannels, E.begin(),
                      LoadILLIndirect::calculateError);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLIndirect.cpp]: Can also use `setHistogram` here?

**Review by SimonHeybrock — COMMENTED** [review]
Looks good now, just a few more improvements:

File: Framework/DataHandling/src/LoadILLIndirect.cpp [hunk 27]
```diff
@@ -291,14 +280,11 @@ void LoadILLIndirect::loadDataIntoTheWorkSpace(
...
-    m_localWorkspace->setSharedX((spec + nb_monitors + i),
-                                 m_localWorkspace->sharedX(0));
-
+	
     // Assign Y
     int *dataSD_p = &dataSD(i, 0, 0);
-    m_localWorkspace->mutableY(spec + nb_monitors + i)
-        .assign(dataSD_p, dataSD_p + m_numberOfChannels);
+	const HistogramData::Counts histoCounts(dataSD_p, dataSD_p + m_numberOfChannels);
+	m_localWorkspace->setHistogram((spec + nb_monitors + i), histoBinEdges, std::move(histoCounts));
```
Comment (SimonHeybrock) [inline Framework/DataHandling/src/LoadILLIndirect.cpp]: Careful, have a look at the original code again: There is no error in this case (for some reason). Your new code auto generates errors so you changed the behavior. To avoid that specify it directly by passing `CountSTandardDeviations(m_numberOfChannels, 0.0)` to `setHistogram` as last argument.

**Comment by SimonHeybrock** [comment]
:shipit:
